### ISSUE TITLE
#[must_use] on a trait method impl does nothing.

### ISSUE THREAD

**Issue by theemathas** [rust-lang/rust-clippy#14202]
### What it does

The lint should warn if `#[must_use]` is applied to a function in an `impl Trait for Type { .... }` block, and suggest applying it to the trait definition instead.

### Advantage

As per [the reference](https://doc.rust-lang.org/stable/reference/attributes/diagnostics.html#r-attributes.diagnostics.must_use.trait-function), `#[must_use]` in the `impl Trait for Type` block does absolutely nothing, and is therefore almost surely a mistake. On the other hand, `#[must_use]` in the trait definition applies to all implementations of the trait.

### Example

```rust
trait Foo {
    fn bar(self) -> i32;
}
impl Foo for () {
    #[must_use]
    fn bar(self) -> i32 {
        1
    }
}
```

Could be written as:

```rust
trait Foo {
    #[must_use]
    fn bar(self) -> i32;
}
impl Foo for () {
    fn bar(self) -> i32 {
        1
    }
}
```

**Comment by y21** [comment]
Maybe this should be an extension to rustc's [`unused_attributes`](https://doc.rust-lang.org/rustc/lints/listing/warn-by-default.html#unused-attributes) lint instead? That one is already used as a general lint for any kind of attribute used in some places where it doesn't do anything.

**Comment by samueltardieu** [comment]
I agree with @y21. I've opened https://github.com/rust-lang/rust/pull/136923 to extend the rustc lint. Closing this one.
